### ISSUE TITLE
Need to do something about comparing numbers to zero

### ISSUE THREAD

**Issue by federicomenaquintero** [mikedilger/float-cmp#13]
In [librsvg] we were getting this test failure on i386, but not on x86_64.  The test checks that a computed transformation matrix is equal to the expected one:

```
thread 'viewport::tests::clip_to_viewport' panicked at 'assertion failed: `(left == right)`
  left: `Some(Matrix { xx: 0.2, yx: 0, xy: 0, yy: 0.2, x0: -0.0000000000000005551115123125783, y0: -0.0000000000000005551115123125783 })`,
 right: `Some(Matrix { xx: 0.2, yx: 0, xy: 0, yy: 0.2, x0: 0, y0: 0 })`', rsvg_internals/src/viewport.rs:199:8
```

The test is a simple `assert_eq!(matrix, expected _matrix)`.  It's obviously wrong; we need to use float-cmp for comparisons or something.

However, we need to compare numbers close to zero sometimes :smile: 

I understand the need for epsilon comparisons.  Currently float-cmp bails out early on zeros.  Maybe we need an ApproxEqEpsilonOrUlps trait?  The function would take both `absolute_diff` and `ulps` arguments.

An alternative for librsvg's need to compare matrices would be for it to transform a point using both, and compare the results with just Ulps.

[librsvg]: https://gitlab.gnome.org/GNOME/librsvg

**Comment by mikedilger** [comment]
Yes, this is a tricky one.

0.0 and -0.0000000000000005551115123125783 are 4844747299143811072 ulps away from each other.   Even if the code didn't bail out early on opposite signs, the ulps method would not work here.

`approx_eq_ratio()` cant work either because we cannot divide by zero.

An epsilon method (which float-cmp does not provide currently) would notice the absolute value of the difference is in fact very small.

Ideally there should be a single function which classifies the input and works in all cases. That also, however, adds to performance overhead if not carefully crafted.

So I like your idea of ApproxEqEpsilonOrUlps.  Did you want to propose a pull request or wait for me to come up with something (it might be a week or so)

**Comment by federicomenaquintero** [comment]
I can start writing a PR early next week.  Now that we have clippy in librsvg's CI, it kills me every time I see the "float equality" warnings, so I'd like to solve this soon :smile: 

Hmm, should I call it ApproxEqMaxDiffOrUlps, per that blog's terminology, and to avoid confusing it with people familar with FLT_EPSILON?

**Comment by mikedilger** [comment]
I'm working on something now.  I'm thinking of calling it just ApproxEq, since I'm hoping to make it cover all cases.  It would take an ulps and an epsilon, and use the epsilon method only if the signs are different or if one of the operands is zero.

**Comment by mikedilger** [comment]
I've pushed something to master.  Please let me know if this looks right.  It is hard to write good code with only two eyes.  It would be especially helpful if you tried it in librsvg and all the warnings went away, I would want to know that.

**Comment by federicomenaquintero** [comment]
This looks good; thank you!  I hope you don't mind some very nitpicky comments; please consider these more as "thinking out loud" than "spotted a bug" :smile: 

Your version has this:
```
    fn approx_eq(&self, other: &f32, ulps: i32, epsilon: f32) -> bool {
        // Compare for exact equality first. This is often true, and so we get the
        // performance benefit of only doing one compare in most cases.
        if *self==*other { return true; }

        // Handle differing signs and zeroes with an epsilon method
        if self.signum() != other.signum() || *self==0.0 || *other==0.0 {
            return (*self - *other).abs() < epsilon;
        }

        let diff: i32 = self.ulps(other);
        diff >= -ulps && diff <= ulps
    }
```

And the version in the blog has this:
```
bool AlmostEqualUlpsAndAbs(float A, float B,
            float maxDiff, int maxUlpsDiff)
{
    // Check if the numbers are really close -- needed
    // when comparing numbers near zero.
    float absDiff = fabs(A - B);
    if (absDiff <= maxDiff)
        return true;
 
    Float_t uA(A);
    Float_t uB(B);
 
    // Different signs means they do not match.
    if (uA.Negative() != uB.Negative())
        return false;
 
    // Find the difference in ULPs.
    int ulpsDiff = abs(uA.i - uB.i);
    if (ulpsDiff <= maxUlpsDiff)
        return true;
 
    return false;
}
```

This is my thought process around "why are they different?".  Your version starts by testing `==`, which works for bit-identical numbers, or comparing +0.0 and -0.0.  Then it does the absolute-difference method if the signs are different, or if one of the numbers is actually zero.  I think the blog's code manages to catch all of those cases by directly doing the absolute-differences method first, without needing to actually check for each special case.  I don't know what's cheaper.

I'll patch librsvg to use this for the problematic tests and report back.  I think I can then replace all our float-comparison code (which does the absolute difference method by hand everywhere) with these traits.

**Comment by federicomenaquintero** [comment]
I've started using this in librsvg and I'm getting some unintuitive results.

A lot of the time I can afford to consider two numbers "equal" if they would have the same representation in Cairo's fixed-point numbers (it uses 24.8 fixnums).  So, I defined a trait that uses yours:

```
use float_cmp::ApproxEq;

const CAIRO_FIXED_FRAC_BITS: u64 = 8;

pub trait ApproxEqCairo: ApproxEq {
    fn approx_eq_cairo(&self, other: &Self) -> bool;
}

impl ApproxEqCairo for f64 {
    fn approx_eq_cairo(&self, other: &f64) -> bool {
        self.approx_eq(other, 1, 1.0 / f64::from(1 << CAIRO_FIXED_FRAC_BITS))
    }
}
```

Shortly after writing that I realized that it is a bit of overkill, as I could simply check that the absolute difference between my numbers is less than Cairo's smallest representable fraction - without going through the Ulps machinery.

Still, instead of changing my impl to do just that, I tried to test it in that state.

This fails; I expected it to work:
```
    #[test]
    fn numbers_approx_equal() {
        assert!(1.0_f64.approx_eq_cairo(&1.001953125_f64)); // 1 + 1/512
    }
```

It fails because a) ApproxEq only runs the absolute-difference method if the numbers differ in signs or either is 0.0; b) I guess I had the expectation of the behavior of the blog you linked, which always runs the differences first, and only tries Ulps if that fails.

The documentation in ApproxEq *does* describe the behavior accurately, so one could argue that I'm using it with the wrong expectations.  However, I'd like to make the case that the blog's scheme is more intuitively useful.  I think it lets one have these assumptions:

* Are you comparing numbers close to zero (but which may be far away in Ulps)?  Fine, just give us your tolerance and we'll use it.
* Are you comparing large numbers, so a small tolerance is meaningless?  Fine, we'll find the distance in Ulps instead.

For librsvg I think I'm going to use the simplified version of just comparing the difference against the minimum fixnum fraction, unless there's a case in the code that requires otherwise.  Still, what do you think of the above?

**Comment by mikedilger** [comment]
I love your analysis.

I was thinking that the epsilon method is too generous for very small numbers.  But now I think that 's wrong, because we are using that epsilon against zeros and differing signs, so anything under the epsilon probably should be seen as negligable.

I was also thinking that the only way floats normally become unequal is due to the processor needing to use a nearest representation, and so the difference in ulps terms should always be quite small, and so the epsilon method is not needed outside of zeroes and differing signs.  I still think this is generally true.

But on balance, and based on the principle of least suprise, you have convinced me.  People will expect any difference below the epsilon they supply to be considered negligable.  Ulps still saves us for very large numbers where one representation is quite a long way off.

And furthermore, if it is still considered wrong by some standard, I will be able to point to the blog and blame him ;-)

In your circumstance, I think you are right, you should just "check that the absolute difference between my numbers is less than Cairo's smallest representable fraction".

**Comment by federicomenaquintero** [comment]
This works perfectly with your latest version; thank you!  I've made it so that

```rust
pub trait ApproxEqCairo: ApproxEq {
    fn approx_eq_cairo(&self, other: &Self) -> bool;
}

impl ApproxEqCairo for f64 {
    fn approx_eq_cairo(&self, other: &f64) -> bool {
        let cairo_smallest_fraction = 1.0 / f64::from(1 << CAIRO_FIXED_FRAC_BITS);
        self.approx_eq(other, cairo_smallest_fraction, 1)
    }
}
```

It works as expected, for the assertion I had before and also for large numbers.

May I ask for a new release of float_cmp to crates.io?  Then I can make librsvg depend on that instead of pointing to your github repo directly; this way "`cargo vendor`" will embed the sources correctly in my tarball :)

**Comment by federicomenaquintero** [comment]
This works as expected; they are 1 ULP away so I'm happy to consider them equal:

```rust
        // 2^53 == 2^53 + 2
        assert!(9_007_199_254_740_992.0.approx_eq_cairo(&9_007_199_254_740_994.0));
```

**Comment by mikedilger** [comment]
Published as 0.4.0.  Thanks for your help.

**Comment by federicomenaquintero** [comment]
Woooo, thanks so much!
